openml/_api_calls.py

```diff
@@ -306,10 +319,16 @@ def __is_checksum_equal(downloaded_file_binary: bytes, md5_checksum: Optional[st
     return md5_checksum == md5_checksum_download
 
 
-def _send_request(request_method, url, data, files=None, md5_checksum=None):
+def _send_request(
+    request_method: str,
+    url: str,
+    data: DATA_TYPE,
+    files: Optional[FILE_ELEMENTS_TYPE] = None,
+    md5_checksum: Optional[str] = None,
+) -> requests.Response:
     n_retries = max(1, config.connection_n_retries)
 
-    response = None
+    response: requests.Response
```

Comment:
Is this for the sake of typing? If yes, is this needed if we know the return type of the function already?

Reply:
Yes. Not doing this will set the type of `response` to `Optional[...]` while it actually cannot be `None`.

Reply:
I am fine with this, but optionally I think it might work if you:
 - `return response` instead of `break`, and
 - use the `else` clause of the `for` loop to raise the last exception
 
 but that's a lot of rewriting, and I think the function probably ought to be refactored in the future anyway (it's kind of unwieldy).